File: src/voice.ts
/**
 * Voice-to-text support for ccbox.
 *
 * Provides push-to-talk voice input using whisper.cpp for speech-to-text.
 * Transcribed text is sent to Claude Code via docker exec or -p flag.
 *
 * Dependency direction:
 *   This module imports from: platform.ts, exec.ts, logger.ts
 *   It should NOT import from: cli, generator, docker-runtime
 */

import { existsSync, mkdirSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

import { exec } from "./exec.js";
import { log } from "./logger.js";
import { detectHostPlatform, commandExists } from "./platform.js";
import { getCcboxTempVoice } from "./constants.js";
import { findRunningContainer } from "./docker-utils.js";
import { getDockerEnv } from "./paths.js";

/** Whisper model sizes with download URLs. */
const WHISPER_MODELS: Record<string, { size: string; url: string }> = {
  base: {
    size: "142MB",
    url: "https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-base.bin",
  },
  small: {
    size: "466MB",
    url: "https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-small.bin",
  },
  tiny: {
    size: "75MB",
    url: "https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-tiny.bin",
  },
};

/**
 * Get the whisper model directory.
 */
function getModelDir(): string {
  return join(homedir(), ".ccbox", "models");
}

/**
 * Get path to whisper model file.
 */
function getModelPath(model = "base"): string {
  return join(getModelDir(), `ggml-${model}.bin`);
}

/**
 * Check if whisper.cpp is available on the host.
 */
export function isWhisperAvailable(): boolean {
  return commandExists("whisper-cpp") || commandExists("whisper") || commandExists("main");
}

/**
 * Get the whisper command name.
 */
function getWhisperCommand(): string | null {
  if (commandExists("whisper-cpp")) { return "whisper-cpp"; }
  if (commandExists("whisper")) { return "whisper"; }
  if (commandExists("main")) { return "main"; }
  return null;
}

/**
 * Download whisper model if not present.
 *
 * @param model - Model name (tiny, base, small).
 * @returns True if model is available.
 */
export async function ensureModel(model = "base"): Promise<boolean> {
  const modelPath = getModelPath(model);
  if (existsSync(modelPath)) {
    return true;
  }

  const modelInfo = WHISPER_MODELS[model];
  if (!modelInfo) {
    log.error(`Unknown whisper model: ${model}. Available: ${Object.keys(WHISPER_MODELS).join(", ")}`);
    return false;
  }

  log.dim(`Downloading whisper ${model} model (${modelInfo.size})...`);
  const modelDir = getModelDir();
  mkdirSync(modelDir, { recursive: true });

  try {
    const result = await exec("curl", [
      "-L", "-o", modelPath, modelInfo.url,
    ], { timeout: 300_000 }); // 5 min for download

    if (result.exitCode === 0 && existsSync(modelPath)) {
      log.success(`Model downloaded: ${model}`);
      return true;
    }
  } catch (e) {
    log.error(`Model download failed: ${String(e)}`);
  }

  return false;
}

/**
 * Record audio from microphone.
 *
 * Uses platform-specific tools for recording.
 *
 * @param durationSec - Maximum recording duration in seconds.
 * @param outputPath - Path to save WAV file.
 * @returns True if recording succeeded.
 */
export async function recordAudio(durationSec: number, outputPath: string): Promise<boolean> {
  const platform = detectHostPlatform();

  try {
    let result;

    switch (platform) {
      case "windows-native":
      case "windows-wsl":
        // Use PowerShell for recording on Windows
        result = await exec("powershell.exe", [
          "-NoProfile", "-Command",
          `Add-Type -AssemblyName System.Speech; $r = New-Object System.Speech.Recognition.SpeechRecognitionEngine; $r.SetInputToDefaultAudioDevice(); Start-Sleep -Seconds ${durationSec}`,
        ], { timeout: (durationSec + 5) * 1000 });
        break;

      case "macos":
        // Use sox (rec) on macOS
        if (commandExists("rec")) {
          result = await exec("rec", [
            outputPath, "trim", "0", String(durationSec),
          ], { timeout: (durationSec + 5) * 1000 });
        } else {
          log.warn("sox not found. Install with: brew install sox");
          return false;
        }
        break;

      case "linux":
        // Use arecord on Linux
        if (commandExists("arecord")) {
          result = await exec("arecord", [
            "-f", "S16_LE", "-r", "16000", "-c", "1",
            "-d", String(durationSec), outputPath,
          ], { timeout: (durationSec + 5) * 1000 });
        } else if (commandExists("rec")) {
          result = await exec("rec", [
            outputPath, "trim", "0", String(durationSec),
          ], { timeout: (durationSec + 5) * 1000 });
        } else {
          log.warn("No audio recorder found. Install arecord (alsa-utils) or sox.");
          return false;
        }
        break;
    }

    return (result?.exitCode ?? 1) === 0;
  } catch (e) {
    log.debug(`Recording failed: ${String(e)}`);
    return false;
  }
}

/**
 * Transcribe audio using whisper.cpp.
 *
 * @param audioPath - Path to WAV file.
 * @param model - Whisper model to use.
 * @returns Transcribed text, or null on failure.
 */
export async function transcribe(audioPath: string, model = "base"): Promise<string | null> {
  const whisperCmd = getWhisperCommand();
  if (!whisperCmd) {
    log.error("whisper.cpp not found in PATH.");
    log.dim("Install: https://github.com/ggerganov/whisper.cpp");
    return null;
  }

  const modelPath = getModelPath(model);
  if (!existsSync(modelPath)) {
    log.error(`Model not found: ${modelPath}`);
    return null;
  }

  try {
    const result = await exec(whisperCmd, [
      "-m", modelPath,
      "-f", audioPath,
      "--no-timestamps",
      "-nt",
    ], { timeout: 60_000 });

    if (result.exitCode === 0) {
      return result.stdout.trim();
    }
  } catch (e) {
    log.debug(`Transcription failed: ${String(e)}`);
  }

  return null;
}

/**
 * Send transcribed text to a running ccbox container.
 *
 * Injects text into the active Claude Code session via ccbox-inject (tmux).
 *
 * @param text - Transcribed text to send.
 * @param containerName - Optional container name.
 * @returns True if sent successfully.
 */
export async function sendToContainer(text: string, containerName?: string): Promise<boolean> {
  const container = containerName ?? await findRunningContainer();
  if (!container) {
    log.error("No running ccbox container found.");
    return false;
  }

  log.dim(`Sending to container: "${text.slice(0, 80)}${text.length > 80 ? "..." : ""}"`);

  try {
    const result = await exec("docker", [
      "exec", container, "/usr/local/bin/ccbox-inject", text,
    ], { timeout: 10_000, env: getDockerEnv() });

    if (result.exitCode === 0) {
      log.success("Voice input sent to active session");
      return true;
    }
    log.error(`Injection failed: ${result.stderr}`);
  } catch (e) {
    log.error(`Failed to send to container: ${String(e)}`);
  }

  return false;
}

/**
 * Run voice-to-text pipeline.
 *
 * Records audio, transcribes with whisper, sends to container.
 *
 * @param options - Voice command options.
 * @returns True if the full pipeline succeeded.
 */
export async function voicePipeline(options: {
  model?: string;
  duration?: number;
  containerName?: string;
} = {}): Promise<boolean> {
  const { model = "base", duration = 10, containerName } = options;

  // Check prerequisites
  if (!isWhisperAvailable()) {
    log.error("whisper.cpp not found in PATH.");
    log.dim("Install: https://github.com/ggerganov/whisper.cpp");
    log.dim("  macOS: brew install whisper-cpp");
    log.dim("  Linux: build from source or use package manager");
    return false;
  }

  // Ensure model
  if (!(await ensureModel(model))) {
    return false;
  }

  // Record audio
  const tmpDir = getCcboxTempVoice();
  mkdirSync(tmpDir, { recursive: true });
  const audioPath = join(tmpDir, `recording-${Date.now()}.wav`);

  log.dim(`Recording (${duration}s max)... Press Ctrl+C to stop.`);
  if (!(await recordAudio(duration, audioPath))) {
    log.warn("Recording failed or no audio captured.");
    return false;
  }

  // Transcribe
  log.dim("Transcribing...");
  const text = await transcribe(audioPath, model);
  if (!text) {
    log.warn("Transcription returned empty result.");
    return false;
  }

  log.dim(`Transcription: "${text}"`);

  // Send to container
  return sendToContainer(text, containerName);
}

File: internal/voice/voice.go
// Package voice provides audio recording and speech-to-text transcription
// using platform audio tools and whisper.cpp.
package voice

import (
	"bytes"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

// Options configures voice recording and transcription.
type Options struct {
	Model    string // Whisper model name (e.g., "base.en", "small", "medium")
	Duration int    // Recording duration in seconds
}

// Pipeline runs the full voice-to-text pipeline: record audio, transcribe, return text.
func Pipeline(opts Options) (string, error) {
	if opts.Duration <= 0 {
		opts.Duration = 10
	}
	if opts.Model == "" {
		opts.Model = "base.en"
	}

	if err := checkDependencies(); err != nil {
		return "", err
	}

	// Ensure whisper model is downloaded before recording
	if _, err := EnsureModel(opts.Model); err != nil {
		return "", fmt.Errorf("model download failed: %w", err)
	}

	audioPath, err := Record(opts.Duration)
	if err != nil {
		return "", fmt.Errorf("recording failed: %w", err)
	}
	defer os.Remove(audioPath)

	text, err := Transcribe(audioPath, opts.Model)
	if err != nil {
		return "", fmt.Errorf("transcription failed: %w", err)
	}

	return strings.TrimSpace(text), nil
}

// Record captures audio from the default microphone for the given duration in seconds.
// Returns the path to the recorded WAV file (16kHz, mono, 16-bit PCM).
func Record(duration int) (string, error) {
	audioFile := filepath.Join(os.TempDir(), fmt.Sprintf("ccbox-voice-%d.wav", time.Now().UnixMilli()))
	durStr := fmt.Sprintf("%d", duration)

	var cmd *exec.Cmd

	switch {
	case commandExists("ffmpeg"):
		cmd = exec.Command("ffmpeg",
			"-y",                    // overwrite output
			"-f", inputFormat(),     // platform-specific input format
			"-i", inputDevice(),     // platform-specific input device
			"-t", durStr,            // duration
			"-ar", "16000",          // 16kHz (whisper requirement)
			"-ac", "1",              // mono
			"-c:a", "pcm_s16le",    // 16-bit PCM
			audioFile,
		)
	case runtime.GOOS == "linux" && commandExists("arecord"):
		cmd = exec.Command("arecord",
			"-f", "S16_LE",
			"-r", "16000",
			"-c", "1",
			"-d", durStr,
			audioFile,
		)
	case commandExists("rec"):
		cmd = exec.Command("rec",
			"-r", "16000",
			"-c", "1",
			"-b", "16",
			audioFile,
			"trim", "0", durStr,
		)
	default:
		return "", fmt.Errorf("no audio recording tool found (install ffmpeg, arecord, or sox)")
	}

	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		os.Remove(audioFile)
		return "", fmt.Errorf("recording command failed: %w", err)
	}

	info, err := os.Stat(audioFile)
	if err != nil || info.Size() == 0 {
		os.Remove(audioFile)
		return "", fmt.Errorf("recording produced no audio data")
	}

	return audioFile, nil
}

// Transcribe runs whisper.cpp on an audio file and returns the transcribed text.
// The model parameter is a whisper model name (e.g., "base.en") which is resolved
// to a file path automatically.
func Transcribe(audioPath, model string) (string, error) {
	modelPath := resolveModelPath(model)

	var cmd *exec.Cmd
	var stdout bytes.Buffer

	switch {
	case commandExists("whisper-cli"):
		cmd = exec.Command("whisper-cli",
			"-m", modelPath,
			"-f", audioPath,
			"--no-timestamps",
		)
	case commandExists("whisper"):
		cmd = exec.Command("whisper",
			"-m", modelPath,
			"-f", audioPath,
			"--no-timestamps",
		)
	default:
		return "", fmt.Errorf("whisper.cpp not found (install whisper-cli from https://github.com/ggerganov/whisper.cpp)")
	}

	cmd.Stdout = &stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("whisper transcription failed: %w", err)
	}

	return stdout.String(), nil
}

// checkDependencies verifies that required external tools are available.
func checkDependencies() error {
	hasRecorder := commandExists("ffmpeg") ||
		(runtime.GOOS == "linux" && commandExists("arecord")) ||
		commandExists("rec")
	if !hasRecorder {
		return fmt.Errorf("no audio recording tool found\nInstall one of: ffmpeg (recommended), arecord (Linux/ALSA), or sox (rec)")
	}

	hasWhisper := commandExists("whisper-cli") || commandExists("whisper")
	if !hasWhisper {
		return fmt.Errorf("whisper.cpp not found\nInstall from: https://github.com/ggerganov/whisper.cpp")
	}

	return nil
}

// EnsureModel checks if the whisper model exists, downloading it if missing.
// Returns the path to the model file.
func EnsureModel(model string) (string, error) {
	modelPath := resolveModelPath(model)
	if _, err := os.Stat(modelPath); err == nil {
		return modelPath, nil
	}

	// Model not found, download it
	url := fmt.Sprintf("https://huggingface.co/ggerganov/whisper.cpp/resolve/main/ggml-%s.bin", model)

	// Ensure cache directory exists
	cacheDir := filepath.Dir(modelPath)
	if err := os.MkdirAll(cacheDir, 0755); err != nil {
		return "", fmt.Errorf("create cache dir: %w", err)
	}

	fmt.Fprintf(os.Stderr, "Downloading whisper model %q...\n", model)

	resp, err := http.Get(url) //nolint:gosec // URL is constructed from known safe base
	if err != nil {
		return "", fmt.Errorf("download model: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("download model: HTTP %d", resp.StatusCode)
	}

	tmpPath := modelPath + ".tmp"
	f, err := os.Create(tmpPath)
	if err != nil {
		return "", fmt.Errorf("create temp file: %w", err)
	}

	if _, err := io.Copy(f, resp.Body); err != nil {
		f.Close()
		os.Remove(tmpPath)
		return "", fmt.Errorf("download model: %w", err)
	}
	f.Close()

	// Atomic rename
	if err := os.Rename(tmpPath, modelPath); err != nil {
		os.Remove(tmpPath)
		return "", fmt.Errorf("install model: %w", err)
	}

	fmt.Fprintf(os.Stderr, "Model %q downloaded to %s\n", model, modelPath)
	return modelPath, nil
}

// resolveModelPath finds the whisper model file on disk.
// Checks common installation locations and returns the first match,
// or the default cache path if none found (whisper-cli will report its own error).
func resolveModelPath(model string) string {
	home := homeDir()
	candidates := []string{
		filepath.Join(home, ".cache", "whisper", "ggml-"+model+".bin"),
		filepath.Join(home, ".local", "share", "whisper", "ggml-"+model+".bin"),
		filepath.Join("models", "ggml-"+model+".bin"),
		model, // allow direct path
	}

	for _, path := range candidates {
		if _, err := os.Stat(path); err == nil {
			return path
		}
	}

	// Return default location; whisper-cli will give a clear error if missing.
	return filepath.Join(home, ".cache", "whisper", "ggml-"+model+".bin")
}

// inputFormat returns the ffmpeg input format flag for the current platform.
func inputFormat() string {
	switch runtime.GOOS {
	case "darwin":
		return "avfoundation"
	case "linux":
		if commandExists("pactl") {
			return "pulse"
		}
		return "alsa"
	case "windows":
		return "dshow"
	default:
		return "pulse"
	}
}

// inputDevice returns the ffmpeg input device identifier for the current platform.
func inputDevice() string {
	switch runtime.GOOS {
	case "darwin":
		return ":0" // default audio input
	case "linux":
		return "default"
	case "windows":
		// Allow override via environment variable for non-standard audio devices
		if dev := os.Getenv("CCBOX_AUDIO_DEVICE"); dev != "" {
			return "audio=" + dev
		}
		return "audio=@device_cm_{33D9A762-90C8-11D0-BD43-00A0C911CE86}\\wave:{0.0.0.00000000}.{default}"
	default:
		return "default"
	}
}

// commandExists checks whether a command is available in the system PATH.
func commandExists(name string) bool {
	_, err := exec.LookPath(name)
	return err == nil
}

// homeDir returns the user's home directory, or "." as fallback.
func homeDir() string {
	home, err := os.UserHomeDir()
	if err != nil {
		return "."
	}
	return home
}
